// Copyright 2024 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

syntax = "proto3";

package xls;

import "xls/ir/op.proto";

// Information about the delay of a single node.
message DelayInfoNodeProto {
  // If in the critical path the total path delay to this node.
  uint64 total_delay_ps = 1;
  // The delay this node contributes.
  uint64 node_delay_ps = 2;
  // The operation this node performs.
  OpProto op = 3;
  // The nodes id-number.
  uint64 id = 4;
  // The IR string of this node.
  string ir = 5;
}

message CriticalPathProto {
  // Total delay of the entire critical path
  uint64 total_delay_ps = 1;
  // The nodes that make up the critical path in order.
  repeated DelayInfoNodeProto nodes = 2;
}

message StageCriticalPathsProto {
  // The critical path of each stage
  map<int64, CriticalPathProto> stage = 1;
}

// Overall delay info
message DelayInfoProto {
  oneof type {
    CriticalPathProto combinational_critical_path = 1;
    StageCriticalPathsProto pipelined_critical_path = 2;
  }
  repeated DelayInfoNodeProto all_nodes = 3;
}
